### PR TITLE
Add support for Multimap of guava with test cases.

### PR DESCRIPTION
**Pull request by Mr14huashao**
Add support for the Multimap interface #284 , but I felt the code structure was a litte overstuffed.
Do you have any suggestions for improvement?
Looking forward to your reply.

### CODE REVIEW

**Comment by dyu** [comment]
@Mr14huashao thank you for creating this PR.  Making guava an optional dependency is preferable. More like a module.
Then the user can then simply register a schema via:
```java
RuntimeSchema.register(ArrayListMultimap.class, MultiMapModule.GenericArrayListMultimap);
RuntimeSchema.map(MultiMap.class, ArrayListMultimap.class);
```

The other way is to apply the changes here on another flavor of protostuff-runtime: `protostuff-runtime-guava` like how `protostuff-runtime-md` is done.

**Comment by Mr14huashao** [comment]
@dyu  Now, the protostuff-runtime-guava module is used.
If you have time, please review code.
Thanks!

File: protostuff-collectionschema/pom.xml [hunk 1]
```diff
@@ -35,6 +35,12 @@
       <artifactId>protostuff-api</artifactId>
       <version>${project.version}</version>
     </dependency>
+    <dependency>
+      <groupId>com.google.guava</groupId>
```
Comment (dyu) [inline protostuff-collectionschema/pom.xml:39]: This belongs in the `protostuff-runtime-guava` module

File: protostuff-runtime-guava/src/main/java/io/protostuff/runtime/RuntimeMultiMapFieldFactory.java [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+import java.lang.reflect.ParameterizedType;
+import java.util.Collection;
+
+/**
+ * Static utility for creating runtime {@link Multimap} fields.
+ *
+ * @author Mr14huashao
+ * @created Otc 09, 2020
+ */
+final class RuntimeMultiMapFieldFactory
```
Comment (dyu) [inline protostuff-runtime-guava/src/main/java/io/protostuff/runtime/RuntimeMultiMapFieldFactory.java:39]: This belongs in the `protostuff-runtime-guava` module

File: protostuff-runtime-registry/src/main/java/io/protostuff/runtime/ExplicitIdStrategy.java [hunk 3]
```diff
@@ -44,6 +44,8 @@
 import io.protostuff.Output;
 import io.protostuff.Pipe;
 import io.protostuff.Schema;
+import com.google.common.collect.Multimap;
+import io.protostuff.MultiMapSchema;
```
Comment (dyu) [inline protostuff-runtime-registry/src/main/java/io/protostuff/runtime/ExplicitIdStrategy.java:48]: Hmm this is a problem. The registry module will have to depend on the `guava` module.
Not sure how we'll obtain adding MultiMap that doesn't have a hard dependency on guava.

File: protostuff-runtime/src/main/java/io/protostuff/runtime/DefaultIdStrategy.java [hunk 4]
```diff
@@ -8,6 +8,7 @@
 import static io.protostuff.runtime.RuntimeFieldFactory.ID_INT32;
 import static io.protostuff.runtime.RuntimeFieldFactory.ID_INT64;
 import static io.protostuff.runtime.RuntimeFieldFactory.ID_SHORT;
+import com.google.common.collect.Multimap;
```
Comment (dyu) [inline protostuff-runtime/src/main/java/io/protostuff/runtime/DefaultIdStrategy.java:11]: The goal is for `protostuff-runtime` to not have a hard dependency on guava.

File: protostuff-runtime-registry/src/main/java/io/protostuff/runtime/ExplicitIdStrategy.java (hunk 3 above)
Comment (Mr14huashao) [reply protostuff-runtime-registry/src/main/java/io/protostuff/runtime/ExplicitIdStrategy.java]: Indeed,  the problem that register depends on guava cannot be resolved for the time being.
The code files related to the `Multimap` are moved to the `protostuff-runtime-guava` module.
Please check, thank you @dyu .

**Review by lastthyme — APPROVED** [review]

**Comment by dyu** [comment]
Applied to `multimap` branch via https://github.com/protostuff/protostuff/pull/339

**Pull request closed without merging**